README
## RETRO FORTH

RETRO is a modern, pragmatic Forth drawing influences from many
sources. It's clean, elegant, tiny, and easy to grasp and adapt
to various uses.

Features:

- Open Source (ISC License)
- Portable (runs on a MISC-style virtual machine)
- Small source & binaries
- Builds into a single, self contained binary for easy deployment
- Sources in literate format, using a Markdown variant

## Quick Start

For most systems (FreeBSD, NetBSD, OpenBSD, macOS, Linux):

    make

You will need a standard C compiler and `make`.

## Executables

Binaries will be placed in the `bin` directory.

The primary executable is `retro`. This is used for running the
examples and the Atua (gopher) & Casket (http) servers that
power forthworks.com.

The `retro` executable embeds the image into the binary, making
it trivial to copy and deploy.

This interface layer also extends the language with many new
words and vocabularies, adds scripting, file i/o, gopher, and
floating point math support.

The `retro` executable can handle a variety of command line
arguments:

    retro -i

Starts the *listener*, a basic REPL for interactive use.

    retro filename

This will run the code in the specified file, then exit. This
is also used to run programs as shell-type scripts using a
header line like `#!/usr/bin/env retro`.

    retro -i -f filename

This will run the code in the specified file, then start the
listener.

    retro -h

Displays a summary of the command line arguments.

Source files for use with `retro` are written with code in
fenced blocks:

    commentary here
     
    ~~~
    code here
    ~~~

Anything outside the fenced blocks will be ignored.

## Documentation

The primary documentation is in RETRO-Book.md (and the formatted
RETRO-Book.html.) Additional notes can be found in the `doc`
directory.

## Commercial Versions

I provide versions of RETRO for iOS and macOS via the Apple app
stores. These use a custom, dual pane editor-based environment
and some platform specific words and extensions. They do use the
same image and virtual machine as `retro`, `retro-repl`, etc.

iOS:   https://itunes.apple.com/us/app/retro-forth-12/id1170943580?ls=1&mt=8

macOS: https://itunes.apple.com/us/app/retro-forth/id1317494014?ls=1&mt=12

## Alternative Implementations

This source tree includes additional implementations in C#,
JavaScript, Pascal, and Python. These are not officially
supported, but are provided for your use if the C
implementations will not suffice.

## Patreon

I have a Patreon at https://www.patreon.com/_crc for those that
want to financially support development. All funds raised via
Pateron are put into development related expenses (server expenses,
app store fees, hardware).

Thanks go out to my current and past patrons:

- Kiyoshi YONEDA
- Rick
- Krinkleneck


CHANGELOG.txt

2017-10-05  Charles Childers  <crc@forthworks.com>

	* Merge branch 'roo-interface'
	* roo finished
	* saving blocks works now
	* s:evaluate no longer generates an error on empty tokens
	* implement insert mode
	* first part of support for modes finished
	* allow for exiting cleanly

2017-10-04  Charles Childers  <crc@forthworks.com>

	* remove old files; revert Makefile changes
	* limit cursor to stay in block; roo.forth now runnable directly
	* more commentary, refine initial keys a little
	* use words instead of a case construct for key handlers
	* cursor movement
	* start some work on block display code
	* begin working on roo (block editor ui)

2017-10-03  Charles Childers  <crc@forthworks.com>

	* add s:evaluate

2017-09-30  Charles Childers  <crc@forthworks.com>

	* remove some debug code from gopher.c

2017-09-29  Charles Childers  <crc@forthworks.com>

	* add gopher:get

2017-09-28  Charles Childers  <crc@forthworks.com>

	* expose files:R+ mode

2017-08-31  Charles Childers  <crc@forthworks.com>

	* update changelog

2017-08-24  Charles Childers  <crc@forthworks.com>

	* add set:reduce
	* next release will be 2017.9
	* add s:copy, document additions to s: namespace

2017-08-21  Charles Childers  <crc@forthworks.com>

	* some new constants in s: namespace
	* bump image to 2017.8
	* some tests for choose, n:negative?

2017-08-02  Charles Childers  <crc@forthworks.com>

	* tests for n:even?, n:odd?, times, n:between?
	* tests for n:min, n:max, n:negate
	* tests for n:limit, v:limit, tuck, fetch-next

2017-07-31  Charles Childers  <crc@forthworks.com>

	* add tests for gt?, lt?, gteq?, lteq?
	* add tests for c:vowel?, c:-vowel?
	* add tests for store, fetch
	* add tests for n:inc, n:dec, const
	* add tests for var, var<n>
	* add tests for v:dec, v:dec-by

2017-07-30  Charles Childers  <crc@forthworks.com>

	* add test for case, s:case
	* initial test for v:inc, v:preserve
	* add initial tests for tri, tri@, tri*
	* add tests for bi, bi@, bi*

2017-07-24  Charles Childers  <crc@forthworks.com>

	* add more tests
	* posix files: more commentary, add file:read-line, file:for-each-line
	* update CHANGELOG
	* tanu is now a filter; closes #12
	* more readability improvements to the Makefile generator
	* use the new build tool to generate a Makefile for the interfaces, extend, and embedimage

2017-07-22  Charles Childers  <crc@forthworks.com>

	* remove unused file
	* remove unused file
	* build.forth: cleaner Makefile output
	* fix listener script
	* new build tool now capable of outputting a valid Makefile
	* more work on new build tool: scan dependencies
	* build.forth: take argument from command line
	* build tool: inching closer
	* more work on build.forth
	* new example

2017-07-21  Charles Childers  <crc@forthworks.com>

	* start work on something like build.c, but in forth
	* begin early work to use new build tool for compiling sources
	* bring in new build tool (intended to help work on issue #13)
	* move documentation into doc dir
	* floating point now an io module
	* close block for floating point code
	* add a .gitattributes
	* seriously raise memory limits: 1M return stack; 4K stack depth; 48M heap
	* split rre.forth better
	* begin work to support embedding source files into rre, other interfaces

2017-07-20  Charles Childers  <crc@forthworks.com>

	* minor update to Primes example; now supports much bigger sets
	* add some more tests
	* Fix example. Closes issue #11
	* begin work on fixing argument handler. see issue #11
	* add a couple more tests

2017-07-19  Charles Childers  <crc@forthworks.com>

	* a few more tests
	* example: using VT100/ANSI escape sequences
	* fix issue w/documentation for s:with-format
	* use .forth suffix for examples, not .md
	* add note on the fenced syntax to Interfaces.md
	* new example: simple implementation of "echo"
	* add support for ~~~ code blocks as well. closes #10
	* begin work on fixing issue #10 (better support for fenced code blocks)
	* Merge pull request #9 from ratanasov/master

2017-07-19  Radoslav Atanasov  <rsrado@gmail.com>

	* Included <stdint.h> to make int32_t visible

2017-07-18  Charles Childers  <crc@forthworks.com>

	* a few more tests
	* new example
	* split legalese into separate LICENSE file
	* add 48 new tests
	* increase test coverage (all of ASCII:, some others)
	* remove some unused code
	* fix initial instance of unspec behaviour in bridge.c (see issue #8)

2017-07-17  Charles Childers  <crc@forthworks.com>

	* start bringing in tests from retro11
	* begin work on an automated testing set
	* document the file: namespace, fix stack comments in rre.forth.md
	* some additional comments on shifts in Nga documentation
	* more work on travis support
	* begin work on travis ci support
	* add sys: namespace, with argc/argv words (addresses issue #6)

2017-07-16  Charles Childers  <crc@forthworks.com>

	* correct f:to-number max value issue; fixes bug #5
	* Merge pull request #4 from ratanasov/master

2017-07-16  Radoslav Atanasov  <rsrado@gmail.com>

	* `@` and `!` added to the prefix table in `RetroForth.md`
	* `make clean` now deletes `bin/repl` too.

2017-07-12  Charles Childers  <crc@forthworks.com>

	* add n:MIN and n:MAX constant
	* fix for issue #3 (bug in n:to-string)

2017-07-11  Charles Childers  <crc@forthworks.com>

	* fix documentation issue
	* update changelog
	* documentation for FPU
	* include f: namespace by default
	* initial checkin supporting FPU (port from WIP on iOS implementation)
	* fix dump-stack

2017-07-08  Charles Childers  <crc@forthworks.com>

	* rename .s (in prep for using . as prefix for floating point)

2017-06-15  Charles Childers  <crc@forthworks.com>

	* add set:length
	* example: finding prime numbers

2017-06-14  Charles Childers  <crc@forthworks.com>

	* stdlib now includes a Retro implementation of muri assembler
	* bump image version
	* update changelog
	* merge in samples, documentation

2017-05-26  Charles Childers  <crc@forthworks.com>

	* documentation updates

tools/muri.c
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <string.h>
#define KiB * 1024
#define MAX_NAMES 1024
#define STRING_LEN 64
#define IMAGE_SIZE 128 KiB
char    Labels[MAX_NAMES][STRING_LEN];
int32_t Pointers[MAX_NAMES];
int32_t np;
char source[1 KiB];
int32_t target[IMAGE_SIZE];
int32_t here;
void save() {
  FILE *fp;
  if ((fp = fopen("ngaImage", "wb")) == NULL) {
    printf("Unable to save the image!\n");
    exit(2);
  }
  fwrite(&target, sizeof(int32_t), here, fp);
  fclose(fp);
}
int32_t lookup(char *name) {
  int32_t slice = -1;
  int32_t n = np;
  while (n > 0) {
    n--;
    if (strcmp(Labels[n], name) == 0)
      slice = Pointers[n];
  }
  return slice;
}
void add_label(char *name, int32_t slice) {
  if (lookup(name) == -1) {
    strcpy(Labels[np], name);
    Pointers[np] = slice;
    np++;
  } else {
    printf("Fatal error: %s already defined\n", name);
    exit(0);
  }
}
void read_line(FILE *file, char *line_buffer) {
  int ch = getc(file);
  int count = 0;
  while ((ch != '\n') && (ch != EOF)) {
    line_buffer[count] = ch;
    count++;
    ch = getc(file);
  }
  line_buffer[count] = '\0';
}
int32_t opcode_for(char *s) {
  if (strcmp(s, "..") == 0) return 0;  if (strcmp(s, "li") == 0) return 1;
  if (strcmp(s, "du") == 0) return 2;  if (strcmp(s, "dr") == 0) return 3;
  if (strcmp(s, "sw") == 0) return 4;  if (strcmp(s, "pu") == 0) return 5;
  if (strcmp(s, "po") == 0) return 6;  if (strcmp(s, "ju") == 0) return 7;
  if (strcmp(s, "ca") == 0) return 8;  if (strcmp(s, "cc") == 0) return 9;
  if (strcmp(s, "re") == 0) return 10; if (strcmp(s, "eq") == 0) return 11;
  if (strcmp(s, "ne") == 0) return 12; if (strcmp(s, "lt") == 0) return 13;
  if (strcmp(s, "gt") == 0) return 14; if (strcmp(s, "fe") == 0) return 15;
  if (strcmp(s, "st") == 0) return 16; if (strcmp(s, "ad") == 0) return 17;
  if (strcmp(s, "su") == 0) return 18; if (strcmp(s, "mu") == 0) return 19;
  if (strcmp(s, "di") == 0) return 20; if (strcmp(s, "an") == 0) return 21;
  if (strcmp(s, "or") == 0) return 22; if (strcmp(s, "xo") == 0) return 23;
  if (strcmp(s, "sh") == 0) return 24; if (strcmp(s, "zr") == 0) return 25;
  if (strcmp(s, "en") == 0) return 26;
  return 0;
}
void pass1(char *fname) {
  int inBlock = 0;
  char *buffer = (char *)source;
  unsigned int opcode;
  char inst[3];
  FILE *fp;
  inst[2] = '\0';
  here = 0;
  fp = fopen(fname, "r");
  if (fp == NULL) {
    printf("Unable to load file\n");
    exit(2);
  }
  while (!feof(fp)) {
    read_line(fp, buffer);
    if (strcmp(buffer, "~~~") == 0) {
      if (inBlock == 0)
        inBlock = 1;
      else
        inBlock = 0;
    } else {
      if (inBlock == 1) {
        if (buffer[1] == '\t' || buffer[1] == ' ') {
          switch (buffer[0]) {
            case 'i': memcpy(inst, buffer + 8, 2);
                      opcode = opcode_for(inst);
                      opcode = opcode << 8;
                      memcpy(inst, buffer + 6, 2);
                      opcode += opcode_for(inst);
                      opcode = opcode << 8;
                      memcpy(inst, buffer + 4, 2);
                      opcode += opcode_for(inst);
                      opcode = opcode << 8;
                      memcpy(inst, buffer + 2, 2);
                      opcode += opcode_for(inst);
                      target[here++] = opcode;
                      break;
            case 'r': target[here++] = -1;
                      break;
            case 'd': target[here++] = atoi(buffer+2);
                      break;
            case 'c': target[here++] = buffer[2];
                      break;
            case 's': opcode = 2;
                      while (opcode < strlen(buffer))
                        target[here++] = buffer[opcode++];
                      target[here++] = 0;
                      break;
            case ':': add_label(buffer+2, here);
                      break;
          }
        }
      }
    }
  }
  fclose(fp);
}
void pass2(char *fname) {
  char *buffer = (char *)source;
  FILE *fp;
  here = 0;
  int inBlock = 0;
  fp = fopen(fname, "r");
  while (!feof(fp)) {
    read_line(fp, buffer);
    if (strcmp(buffer, "~~~") == 0) {
      if (inBlock == 0)
        inBlock = 1;
      else
        inBlock = 0;
    } else {
      if (inBlock == 1) {
        switch (buffer[0]) {
          case 'i': here++;                             break;
          case 'r': target[here++] = lookup(buffer+2);
                    if (lookup(buffer+2) == -1)
                      printf("Lookup failed: '%s'\n", buffer+2);
                                                        break;
          case 'd': here++;                             break;
          case 'c': here++;                             break;
          case 's': here = here + strlen(buffer) - 1;   break;
          case ':':                                     break;
        }
      }
    }
  }
  fclose(fp);
}
int main(int argc, char **argv) {
  np = 0;
  if (argc > 1) {
    pass1(argv[1]);
    pass2(argv[1]);
    save();
    printf("Wrote %d cells to ngaImage\n", here);
  }
  else
    printf("muri\n(c) 2017 charles childers\n\n%s filename\n", argv[0]);
  return 0;
}
